File: openai/openai_server.py
#!/usr/bin/env python3
"""FireRedASR-AED OpenAI 兼容服务（/v1/audio/transcriptions）。"""
import argparse
import json
import re
import tempfile
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from firered_asr import FireredASR


def parse_multipart(content_type, body):
    m = re.search(r'boundary=(?:"([^"]+)"|([^;]+))', content_type or "")
    boundary = (m.group(1) or m.group(2)).strip() if m else None
    if not boundary:
        return None
    delim = b"--" + boundary.encode()
    parts = body.split(delim)
    files = {}
    for part in parts:
        if b"\r\n\r\n" not in part:
            continue
        header, _, data = part.partition(b"\r\n\r\n")
        if data.endswith(b"\r\n"):
            data = data[:-2]
        name = None
        filename = None
        for line in header.split(b"\r\n"):
            if line.lower().startswith(b"content-disposition:"):
                nm = re.search(rb'name="([^"]+)"', line, re.I)
                fn = re.search(rb'filename="([^"]*)"', line, re.I)
                if nm:
                    name = nm.group(1).decode()
                if fn:
                    filename = fn.group(1).decode()
        if name:
            files[name] = {"content": data, "filename": filename}
    return files


def make_handler(asr):
    class Handler(BaseHTTPRequestHandler):
        def _send_json(self, obj, code=200):
            body = json.dumps(obj, ensure_ascii=False).encode()
            self.send_response(code)
            self.send_header("Content-Type", "application/json; charset=utf-8")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

        def do_GET(self):
            if self.path.rstrip("/") == "/v1/models":
                self._send_json({"object": "list", "data": [
                    {"id": "fireredasr-aed", "object": "model", "owned_by": "axera"}]})
            else:
                self._send_json({"error": "not found"}, 404)

        def do_POST(self):
            if self.path.rstrip("/") != "/v1/audio/transcriptions":
                self._send_json({"error": "not found"}, 404)
                return
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)
            files = parse_multipart(self.headers.get("Content-Type", ""), body)
            if not files or "file" not in files:
                self._send_json({"error": "missing multipart field 'file'"}, 400)
                return
            data = files["file"]["content"]
            with tempfile.NamedTemporaryFile(suffix=".wav", delete=False) as f:
                f.write(data)
                tmp = f.name
            try:
                text = asr.transcribe(tmp)
                self._send_json({"text": text})
            except Exception as e:
                self._send_json({"error": str(e)}, 500)

        def log_message(self, *args):
            pass
    return Handler


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--encoder", required=True)
    ap.add_argument("--decoder", required=True)
    ap.add_argument("--fsmn-vad", required=True)
    ap.add_argument("--fsmn-cmvn", required=True)
    ap.add_argument("--cmvn", required=True)
    ap.add_argument("--dict", required=True)
    ap.add_argument("--pe", required=True)
    ap.add_argument("--max-dur", type=int, default=10)
    ap.add_argument("--max-steps", type=int, default=128)
    ap.add_argument("--host", default="0.0.0.0")
    ap.add_argument("--port", type=int, default=8000)
    args = ap.parse_args()

    asr = FireredASR(args.encoder, args.decoder, args.fsmn_vad, args.fsmn_cmvn,
                     args.cmvn, args.dict, args.pe,
                     args.max_dur, args.max_steps)
    print(f"OpenAI API listening on {args.host}:{args.port}")
    ThreadingHTTPServer((args.host, args.port), make_handler(asr)).serve_forever()


if __name__ == "__main__":
    main()

File: cpp/openai_server.cpp
// FireRedASR-AED OpenAI 兼容 API 服务（AX650N）
#include <cstdio>
#include <cstring>
#include <string>

#include "httplib.h"
#include "sdk_lib.hpp"

static std::string json_escape(const std::string& s) {
    std::string o;
    for (char c : s) {
        if (c == '"' || c == '\\') { o += '\\'; o += c; }
        else if (c == '\n') o += "\\n";
        else if (c == '\r') o += "\\r";
        else if (c == '\t') o += "\\t";
        else o += c;
    }
    return o;
}

int main(int argc, char** argv) {
    setvbuf(stdout, nullptr, _IONBF, 0);
    if (argc < 10) {
        printf("usage: %s <encoder> <decoder> <fsmn_vad> <fsmn_cmvn> "
               "<cmvn> <dict> <pe.bin> <max_dur_s> <max_steps> "
               "[port] [vad_min_speech_ms]\n", argv[0]);
        return 1;
    }
    const char* enc = argv[1];
    const char* dec = argv[2];
    const char* fsmn_vad = argv[3];
    const char* fsmn_cmvn = argv[4];
    const char* cmvn = argv[5];
    const char* dict = argv[6];
    const char* pe = argv[7];
    int max_dur = atoi(argv[8]);
    int max_steps = atoi(argv[9]);
    int port = argc > 10 ? atoi(argv[10]) : 8000;
    int fsmn_min_ms = argc > 11 ? atoi(argv[11]) : 1000;

    FireredSdk sdk;
    if (!sdk.Init(enc, dec, fsmn_vad, fsmn_cmvn,
                  cmvn, dict, pe, max_dur, max_steps, fsmn_min_ms)) {
        printf("SDK init failed\n");
        return 1;
    }

    httplib::Server svr;
    svr.Get("/v1/models", [](const httplib::Request&, httplib::Response& res) {
        res.set_content(
            "{\"object\":\"list\",\"data\":[{\"id\":\"fireredasr-aed\","
            "\"object\":\"model\",\"owned_by\":\"axera\"}]}",
            "application/json");
    });
    svr.Post("/v1/audio/transcriptions", [&](const httplib::Request& req, httplib::Response& res) {
        if (!req.form.has_file("file")) {
            res.status = 400;
            res.set_content("{\"error\":\"missing multipart field 'file'\"}", "application/json");
            return;
        }
        auto f = req.form.get_file("file");
        const std::string tmp = "/tmp/firered_openai_input.wav";
        FILE* fp = fopen(tmp.c_str(), "wb");
        if (!fp) {
            res.status = 500;
            res.set_content("{\"error\":\"cannot write temp wav\"}", "application/json");
            return;
        }
        fwrite(f.content.data(), 1, f.content.size(), fp);
        fclose(fp);
        double vad_ms = 0, enc_ms = 0, dec_ms = 0;
        std::string text = sdk.Transcribe(tmp.c_str(), &vad_ms, &enc_ms, &dec_ms);
        std::string body = "{\"text\":\"" + json_escape(text) + "\"}";
        res.set_content(body, "application/json");
    });
    printf("OpenAI API listening on 0.0.0.0:%d\n", port);
    svr.listen("0.0.0.0", port);
    return 0;
}
